### ISSUE TITLE
Logs pipeline operators logic assume UTC on kubernetes node, can cause incorrect timestamp in Splunk

### ISSUE THREAD

**Issue by matthewmodestino** [signalfx/splunk-otel-collector-chart#644]
Hey Team!

Was working with a colleague (shout out to Mike Sakahara!) and discovered that our logs pipeline assumes kubernetes host will be in UTC. While this is definitely the most common scenario, and we do encourage users to always use UTC, if the host is not set to UTC, it breaks our hardcoded parsing logic:

First, if the container engine is not set in the helm chart (which i also always encourage) the `get-format` operator assumes Zulu timestamp:

https://github.com/signalfx/splunk-otel-collector-chart/blob/f9abebe4814f6fbc3988f009db0772e76ac1fe65/rendered/manifests/otel-logs/configmap-agent.yaml#L149-L157

Then when passed to the matched format parser (which actually may not even match anymore) the log format parser also fails to extract the timestamp due to hardcoded look for Zulu:

https://github.com/signalfx/splunk-otel-collector-chart/blob/f9abebe4814f6fbc3988f009db0772e76ac1fe65/rendered/manifests/otel-logs/configmap-agent.yaml#L172-L177

Ideally we should improve the regex to be able to handle the chance the host timestamp may not be UTC/Z and parse out the timestamp correctly. 

Here is an example of containerd logs on host set to GMT-6

```
2023-01-23T12:09:28.344779078-06:00 stdout F {"caller":"service_controller.go:103","controller":"ServiceReconciler","end reconcile":"kube-system/microk8s.io-hostpath","level":"info","ts":"2023-01-23T18:09:28Z"}
2023-01-23T12:09:28.344779078-06:00 stdout F {"caller":"service_controller.go:103","controller":"ServiceReconciler","end reconcile":"kube-system/microk8s.io-hostpath","level":"info","ts":"2023-01-23T18:09:28Z"}
root@zach:/var/log/pods/metallb-system_speaker-z78rf_57df697a-e63d-48d5-bccb-562074f1e6c9/speaker# date
Tue Jan 24 09:25:24 AM CST 2023
```
checked our regex parser in regex101 golang and it does not match. likely breaks our parsing completely, tho events still arrive in Splunk, but the timestamp is incorrectly showing 3 AM, when event is from 3 pm. 

<img width="1296" alt="image" src="https://user-images.githubusercontent.com/31549209/214347368-f11bdbd4-b480-4898-8c2c-f7c5cc3dd8c4.png">

**Comment by matthewmodestino** [comment]
Ran into this again today when setting `containerRuntime: containerd` on a MicroK8s install running on a host in CST. 

The tell tail sign is if you see the following errors in the filelog operators:

```
2023-03-16T16:12:14.409Z    error   recombine/recombine.go:301  entry does not contain the combine_field    {"kind": "receiver", "name": "filelog", "data_type": "logs", "operator_id": "containerd-recombine", "operator_type": "recombine"}
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/transformer/recombine.(*Transformer).addToBatch
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/transformer/recombine/recombine.go:301
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/transformer/recombine.(*Transformer).Process
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/transformer/recombine/recombine.go:266
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*WriterOperator).Write
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/writer.go:64
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*TransformerOperator).HandleEntryError
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/transformer.go:112
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ParseWith
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:151
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWithCallback
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:123
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWith
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:109
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/regex.(*Parser).Process
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/parser/regex/regex.go:110
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*WriterOperator).Write
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/writer.go:64
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/file.(*Input).emit
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/input/file/file.go:65
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Reader).ReadToEnd
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/fileconsumer/reader.go:87
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).consume.func1
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/fileconsumer/file.go:151
2023-03-16T16:12:14.409Z    error   helper/transformer.go:110   Failed to process entry {"kind": "receiver", "name": "filelog", "data_type": "logs", "operator_id": "parser-containerd", "operator_type": "regex_parser", "error": "regex pattern does not match", "action": "send", "entry": {"observed_timestamp":"2023-03-16T16:12:14.409398702Z","timestamp":"0001-01-01T00:00:00Z","body":"2023-03-16T11:12:13.522733476-05:00 stderr F 2023/03/16 16:12:13 settings.go:331: Set config to [/conf/relay.yaml]","attributes":{"log.file.path":"/var/log/pods/otel_joe-tko-otel-splunk-otel-collector-agent-pvx62_209b8033-e541-4811-8c8c-efe0473179e9/otel-collector/0.log"},"severity":0,"scope_name":""}}
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*TransformerOperator).HandleEntryError
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/transformer.go:110
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ParseWith
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:151
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWithCallback
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:123
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*ParserOperator).ProcessWith
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/parser.go:109
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/parser/regex.(*Parser).Process
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/parser/regex/regex.go:110
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper.(*WriterOperator).Write
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/helper/writer.go:64
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/input/file.(*Input).emit
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/operator/input/file/file.go:65
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Reader).ReadToEnd
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/fileconsumer/reader.go:87
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer.(*Manager).consume.func1
    github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.72.0/fileconsumer/file.go:151
2023-03-16T16:12:23.918Z    error   helper/transformer.go:110   Failed to process entry {"kind": "receiver", "name": "filelog", "data_type": "logs", "operator_id": "move5", "operator_type": "move", "error": "move: field does not exist", "action": "send", "entry": {"observed_timestamp":"2023-03-16T16:12:14.409613859Z","timestamp":"0001-01-01T00:00:00Z","body":"2023-03-16T11:12:10.108240126-05:00 stderr F 2023/03/16 16:12:10 Otel checkpoint already present. no need to migrate.","attributes":{"log":"","log.file.path":"/var/log/pods/otel_joe-tko-otel-splunk-otel-collector-agent-pvx62_209b8033-e541-4811-8c8c-efe0473179e9/migrate-checkpoint/0.log"},"resource":{"com.splunk.sourcetype":"kube:container:migrate-checkpoint","k8s.container.name":"migrate-checkpoint","k8s.container.restart_count":"0","k8s.namespace.name":"otel","k8s.pod.name":"joe-tko-otel-splunk-otel-collector-agent-pvx62","k8s.pod.uid":"209b8033-e541-4811-8c8c-efe0473179e9"},"severity":0,"scope_name":""}}
```

When i first posted it was only causing timestamp issues, in 0.72.0 its breaking the entire logging operator logic and caused events not to be indexed. 

The `cri-o` setting for `containerRuntime` was able to parse correctly, so workaround was either not setting the `containerRuntime` at all, or setting `cri-o`. 

`containerd` parser needs to be improved to not be so brittle against timezone formats.

**Comment by matthewmodestino** [comment]
Bump. Need this fixed. Seeing more users who have their k8s node set to a timezone other than UTC.

This causes the `containerd` filelog regex parser to not match and no logs make it through the pipeline.

Working on a override workaround.

**Comment by VihasMakwana** [comment]
maybe we can add a field in values.yaml to hold the timestamp format?

**Comment by matthewmodestino** [comment]
I was able to replicate and workaround like this:

On Ubuntu 20.04, used `timedatectl` to set my node time to `America/Toronto`

Then installed MicroK8s and verified with `kubectl logs --timestamps=true` that the k8s logs had -4:00 timestamp

`2023-06-13T09:33:03.082606034-04:00 2023/06/13 13:33:03 settings.go:373: Set config to [/conf/relay.yaml]`

I then used the `config` section of the agent to override the operators...had to include them all to ensure none were lost in the override. I updated the regex to not anchor on `Z` and used gotime layout. 

```
 config:
    receivers:
      filelog:
        operators:
        - id: parser-containerd
          regex: ^(?P<time>[^ ]+) (?P<stream>stdout|stderr) (?P<logtag>[^ ]*) ?(?P<log>.*)$
          timestamp:
            layout: "2006-01-02T15:04:05.999999999-07:00"
            layout_type: gotime
            parse_from: attributes.time
          type: regex_parser
        - combine_field: attributes.log
          combine_with: ""
          id: containerd-recombine
          is_last_entry: attributes.logtag == 'F'
          max_log_size: 1048576
          output: handle_empty_log
          source_identifier: attributes["log.file.path"]
          type: recombine
        - field: attributes.log
          id: handle_empty_log
          if: attributes.log == nil
          type: add
          value: ""
        - parse_from: attributes["log.file.path"]
          regex: ^\/var\/log\/pods\/(?P<namespace>[^_]+)_(?P<pod_name>[^_]+)_(?P<uid>[^\/]+)\/(?P<container_name>[^\._]+)\/(?P<restart_count>\d+)\.log$
          type: regex_parser
        - from: attributes.uid
          to: resource["k8s.pod.uid"]
          type: move
        - from: attributes.restart_count
          to: resource["k8s.container.restart_count"]
          type: move
        - from: attributes.container_name
          to: resource["k8s.container.name"]
          type: move
        - from: attributes.namespace
          to: resource["k8s.namespace.name"]
          type: move
        - from: attributes.pod_name
          to: resource["k8s.pod.name"]
          type: move
        - field: resource["com.splunk.sourcetype"]
          type: add
          value: EXPR("kube:container:"+resource["k8s.container.name"])
        - from: attributes.stream
          to: attributes["log.iostream"]
          type: move
        - from: attributes["log.file.path"]
          to: resource["com.splunk.source"]
          type: move
        - from: attributes.log
          id: clean-up-log-record
          to: body
          type: move

```
looks like its working now...

<img width="1399" alt="image" src="https://github.com/signalfx/splunk-otel-collector-chart/assets/31549209/e2e90e1f-987c-4213-b2a4-cc0d8c94aa74">


tldr, please update the regex and timestamp parsing for the containerd pipeline

**Comment by VihasMakwana** [comment]
this surely is not fun.
will update values.yaml with configurable timestamp format and test it out.

**Comment by jvoravong** [comment]
Created a PR with a potential fix. The PR links to the ticket the community created which I think is a duplicate of this issue.

**Comment by jvoravong** [comment]
@matthewmodestino the crio-fix was released with v0.79.1 of the chart. Can you confirm if this fixes this issue?

**Comment by matthewmodestino** [comment]
To be clear, it was containerd parsing that mostly triggered the issue in practice. Will check the configs and try it out.

**Comment by VihasMakwana** [comment]
I can still reproduce this issue in the latest release.
@jvoravong this issue occurs when we set `containerRuntime: containerd`. if we leave it blank, it would work just fine.

**Comment by matthewmodestino** [comment]
Even if containerRuntime is left blank, i still believe its a problem as the regexes dont match in the operators. At the end of the day the containerd and cri-o parsers should not anchor on "Z" in the regex. The regex needs to account for Z not being the end of the timestamp.

**Comment by matthewmodestino** [comment]
Bump. Still receiving reports of folks running into timezone parsing for containerd when node is not in UTC

**Comment by VihasMakwana** [comment]
@matthewmodestino my PR will fix it, but it comes with a breaking change as we remove the containerRuntime config.
I'm currently testing the changes on EKS and GCP.

**Comment by matthewmodestino** [comment]
Ok. Pretty sure all this needed was a better regex, no??

**Comment by VihasMakwana** [comment]
Okay, I'm actually working on same right now.
This breaking change can work later. Let me roll out a PR just for this fix and not worry about breaking anything. 
give me couple of hours

**Comment by jvoravong** [comment]
It would be useful to start tracking the versions of Kubernetes and containerd affected by this issue, started using a command like [this](https://github.com/signalfx/splunk-otel-collector-chart/blob/a0325d6bb0d21eec3289f8e710005d324420c950/.github/workflows/functional_test.yaml#L62C1-L63C1).

`echo "Kubernetes $(kubectl version --short | grep -E 'Client|Server')" && echo "Container Runtime for Node: $(kubectl get node -o=jsonpath='{.items[0].metadata.name}'): $(kubectl get node -o=jsonpath='{.items[0].status.nodeInfo.containerRuntimeVersion}')"`

**Comment by VihasMakwana** [comment]
@matthewmodestino can you test it out now?

**Comment by matthewmodestino** [comment]
Sorry, missed this. Do i just need to point to a different image? 

Let me know what steps to take to test the latest pull

**Comment by atoulme** [comment]
Closing as inactive. Please continue the conversation on a Splunk idea if you want to enhance further.
